proptest/src/num/float_samplers.rs

```diff
@@ -0,0 +1,338 @@
+
+                    FloatUniform {
+                        uniform: Uniform::new(0, values.count),
+                        values,
+                    }
+                }
+
+                fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
+                    $wrapper(self.values.get(self.uniform.sample(rng)))
+                }
+            }
+
+            impl SampleUniform for $wrapper {
+                type Sampler = FloatUniform;
+            }
+
+            #[derive(Clone, Copy, Debug)]
+            struct SampleValueCollection {
+                start: $typ,
+                end: $typ,
+                step: $typ,
+                count: $int_typ,
+            }
+
+            // Values greater than MAX_PRECISE_INT may be rounded when converted to float.
+            const MAX_PRECISE_INT: $int_typ =
+                (2 as $int_typ).pow($typ::MANTISSA_DIGITS);
+
+            // The collection of sample values that may be generated by UniformF32U.
+            impl SampleValueCollection {
+                fn new_inclusive(low: $typ, high: $typ) -> Self {
+                    assert!(low.is_finite(), "low finite");
+                    assert!(high.is_finite(), "high finite");
+                    assert!(high - low >= 0., "invalid range");
+
+                    let min_abs = $typ::min(low.abs(), high.abs());
+                    let max_abs = $typ::max(low.abs(), high.abs());
+
+                    let gap = last_gap(max_abs);
+
+                    let (start, end, step) = if low.abs() < high.abs() {
+                        (high, low, -gap)
+                    } else {
+                        (low, high, gap)
+                    };
+
+                    let min_gaps = min_abs / gap;
+                    let max_gaps = max_abs / gap;
+                    debug_assert!(
+                        max_gaps.floor() == max_gaps,
+                        "max_gaps is an integer"
+                    );
+
+                    let count = if low.signum() == high.signum() {
+                        max_gaps as $int_typ - min_gaps.floor() as $int_typ
+                    } else {
+                        max_gaps as $int_typ + min_gaps.ceil() as $int_typ
+                    } + 1;
+                    debug_assert!(count - 1 <= 2 * MAX_PRECISE_INT);
+
+                    Self {
+                        start,
+                        end,
+                        step,
+                        count,
+                    }
+                }
+
+                fn get(&self, index: $int_typ) -> $typ {
+                    assert!(index < self.count, "index out of bounds");
+
+                    if index == self.count - 1 {
+                        return self.end;
+                    }
+
+                    ((index / 2) as $typ).mul_add(
+                        2. * self.step,
+                        (index % 2) as $typ * self.step + self.start,
+                    )
```

Comment:
I think this little section of code would benefit from a comment, very unclear to me what it's intent is. Not sure why we'd divide by 2 only to then multiply but 2. I understand that floats are finicky, so maybe you're exploiting something that should be added to the comment.

Reply:
This is clear now, thank you